Add landing page when deep linking to native app

dmeza opened the pull request:
#### Summary
These changes add a new route `vault` to a landing page that allows to fire the protocol to native client if installed or give the user the option to download the native app or redirects to the webapp after 5 seconds.

This is an example if the native app is installed and the protocol is successfully registered:
<img width="1212" alt="screen shot 2017-10-26 at 11 31 06 pm" src="https://user-images.githubusercontent.com/160621/32088231-0b73c2a6-bae2-11e7-9ecc-7ff87b3abfbf.png">

#### Ticket Link
https://github.com/mattermost/desktop/issues/628

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [ ] Added or updated unit tests (required for all new features)
- [ ] Has server changes (please link)
- [ ] Has redux changes (please link)
- [x] Has UI changes
- [ ] Includes text changes and localization file
- [ ] Touches critical sections of the codebase (auth, posting, etc.)

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0c5a43f94b0eba8db.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0c5a43f94b0eba8db

jasonblais commented:
@dmeza Can you help rebase your branch and resolve the build failures?

dmeza commented:
@jasonblais rebased and fixed build failures.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-03689a58ed807bb97.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-03689a58ed807bb97

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0b65410b254380611.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0b65410b254380611

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-009e0a4b1e312a6c3.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-009e0a4b1e312a6c3

dmeza commented:
@jasonblais rebase to the latest.

jasonblais commented:
Sorry @dmeza for the delay in reviewing this PR. Have asked one of our team members to help take a look

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-076621ee81a0afee2.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-076621ee81a0afee2

dmeza commented:
@esethna rebased to the latest from master and tested but I couldn't reproduce the issue. I created many users and tested with a couple of teams, but did not get the `undefined` issue. Please let me know if you have any special configuration or steps to reproduce. Thank you.

jasonblais commented:
FYI, added a ticket in JIRA just so it can be tracked by the platform team as well: https://mattermost.atlassian.net/browse/PLT-8231

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

herooftimeandspace commented:
Just to make sure, are you on OS X High Sierra with Chrome? We noticed a
bug in Chromium that caused blank tabs to load from ANY link clicked within
uChat or other applications. The fix is to restart Chrome, which installs
the pending updates that cause the issue. Bug should be fixed in the latest
build of Chrome.

Link Campbell
Interim Product Manager | uChat
TS Dev | 555 Market Street, San Francisco, CA 94103

On Mon, Nov 27, 2017 at 1:24 PM, Eric Sethna <notifications@github.com>
wrote:

> *@esethna* requested changes on this pull request.
>
> @dmeza <https://github.com/dmeza> Looks like rebasing from master solved
> the issue I was having above ^.
>
> I'm having trouble triggering the landing page though, I just seem to get
> a blank page when I click the mattermost://... link in the spinmint from
> a computer without the desktop app installed. See the messages posted in
> the spinmint Town Square, I am testing by:
>
>    1. Posting a message and getting its permalink.
>    2. Post the permalink replacing http with mattermost
>    3. Click on the link from a computer without desktop app installed
>
> [image: image]
> <https://user-images.githubusercontent.com/14333569/33290014-9a619484-d37e-11e7-8427-b081bccb5c19.png>
>
> Can you help provide me with updated testing instructions to view the
> landing page if this is incorrect ^.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mattermost/mattermost-webapp/pull/208#pullrequestreview-79316835>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AfEwjsFqiHq0YhAxULYAMkcO2TCZVYYbks5s6yiFgaJpZM4QIiln>
> .
>

esethna commented:
Thanks @herooftimeandspace, I am on Chrome 62.0.3202.94 (up to date) on High Sierra. I get a new blank tab when opening any mattermost:// permalinks posted in the test server.

I also tried on my Windows 10 machine after uninstalling the desktop app, I got the same blank tab after accepting the "Open URL:mattermost request":
![mattermost url](https://user-images.githubusercontent.com/14333569/33459142-20a3f2a8-d5e6-11e7-87d3-aa3927dbd186.gif)

cc// @dmeza

herooftimeandspace commented:
Ok, thanks. I wanted to confirm that it was an issue with the protocol and
not an issues with Chromium. Thank you for your time.

Link Campbell
Interim Product Manager | uChat
TS Dev | 555 Market Street, San Francisco, CA 94103

On Thu, Nov 30, 2017 at 2:52 PM, Eric Sethna <notifications@github.com>
wrote:

> Thanks @herooftimeandspace <https://github.com/herooftimeandspace>, I am
> on Chrome 62.0.3202.94 (up to date) on High Sierra. I get a new blank tab
> when opening any mattermost:// permalinks posted in the test server.
>
> I also tried on my Windows 10 machine after uninstalling the desktop app,
> I got the same blank tab after accepting the "Open URL:mattermost request":
> [image: mattermost url]
> <https://user-images.githubusercontent.com/14333569/33459142-20a3f2a8-d5e6-11e7-87d3-aa3927dbd186.gif>
>
> cc// @dmeza <https://github.com/dmeza>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mattermost/mattermost-webapp/pull/208#issuecomment-348347110>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AfEwjsPTRxKjWFmR8SgpW20A1G-DEJXcks5s7zGvgaJpZM4QIiln>
> .
>

esethna commented:
@dmeza let me know if I can help at all to debug this?

dmeza commented:
Hi @esethna and @csduarte I'm blocked with this PR. I'm able to reproduce the blank page, but can't find how to fix it. If you have some ideas I'm open to try them.
I followed this tutorial on the original implementation: https://developers.tune.com/sdk/deep-linking-to-your-mobile-app-from-your-website/

The landing page is basically for desktop. For the React native apps the changes are done to not need the landing page. 

For desktop this is the line that registers the protocol: https://github.com/mattermost/desktop/blob/master/src/main.js#L350
This is the electron documentation for registering the protocol: https://electronjs.org/docs/api/protocol

miguelespinoza commented:
@dmeza, Here's the commit with the protocol detection.  Turns out when changing url from http | https to 'mattermost' if the protocol is not supported it throws an error.

Unfortunately, there's no easy/clean approach to this since each browser is different. `custom-protocol-detection` is a library that supports solutions for popular browsers (Chrome, IE, Edge, Safari, Opera).

Using this solution requires to change the strategy for 3 scenarios:
1. protocol is supported
    * window pop-up to `Open Mattermost`
    * webpage redirects to web app
2. protocol is not supported

    * display message: "Unable to open Mattermost"
<img width="590" alt="screen shot 2017-12-26 at 3 35 11 pm" src="https://user-images.githubusercontent.com/6182543/34364943-fb05ee3e-ea58-11e7-92e5-5578a820be96.png">
3. obscure browser is not supported by the `custom-protocol-detection` 

    * display message: "This browser does not support opening applications."

<img width="783" alt="screen shot 2017-12-26 at 3 34 36 pm" src="https://user-images.githubusercontent.com/6182543/34364951-1b424148-ea59-11e7-8522-aa192596580d.png">

dmeza commented:
@esethna tested the changes from @miguelespinoza. It works great on Chrome, FF and IE. I couldn't get it to work on Safari or Opera. Open to other ideas.

jasonblais commented:
@dmeza 

> The landing page is basically for desktop. For the React native apps the changes are done to not need the landing page.

I'm curious what those changes were and whether any of them could be applied to desktop?

> It works great on Chrome, FF and IE. I couldn't get it to work on Safari or Opera. Open to other ideas.

What happens on Safari and Opera?

dmeza commented:
@jasonblais @csduarte:
@jarredwitt is the one that implemented that for react native apps. I don't think it's possible to do it in the same way for the desktop app.
@miguelespinoza not sure if you have any ideas on how to make it work for Safari and Opera?
Jason anyone else is welcome to chip in.

jasonblais commented:
@dmeza 

> @miguelespinoza not sure if you have any ideas on how to make it work for Safari and Opera?
Jason anyone else is welcome to chip in.

I was more curious on what currently happens on Safari or Opera? Depending on what the experience is, it might be okay given those browsers are used less than Chrome/IE11/Firefox.

dmeza commented:
@jasonblais:
> I was more curious on what currently happens on Safari or Opera? Depending on what the experience is, it might be okay given those browsers are used less than Chrome/IE11/Firefox.
@miguelespinoza reported that Safari and Opera were working. Testing on my machine, Safari and Opera did not redirect to the protocol or open a confirmation modal when desktop app was installed.

dmeza commented:
@jasonblais rebased to latest from master so you can test it if you'd like.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0c3f84681f900f54b.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0c3f84681f900f54b

jasonblais commented:
I did a quick smoke test on Mac Chrome/Safari and it works for me when the app is installed.

I'll let @esethna to help with more thorough testing though, given he did them before (@esethna let me know if I can help test further, would be happy to).

@esethna We should also get Mike G. or Asaad to review the UI, and make changes if necessary. I think some of the spacing and font sizes may be inconsistent with our other pages. Propose we make the changes ourselves so we don't take Shift's time.

Nice work @dmeza with this.

dmeza commented:
Thank you @jasonblais. It was @miguelespinoza's  work that unblocked the white screen.

esethna commented:
Thanks JB,

FYI I've asked help from @asaadmahmood to help make any formatting tweaks necessary to the page.

asaadmahmood commented:
@dmeza Can you help have a look at the build issue?

dmeza commented:
@asaadmahmood @esethna rebased to latest from master and uncommented safeOpenProtocol() for the full functionality to work. This is a screen capture of the style changes I see locally:
<img width="1143" alt="screen shot 2018-01-17 at 10 16 53 am" src="https://user-images.githubusercontent.com/160621/35050298-7a3e7680-fb70-11e7-93b2-ec02fca6739b.png">

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0e8ffc91dcbf6e068.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0e8ffc91dcbf6e068

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0dc770926d96aa84a.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0dc770926d96aa84a

esethna commented:
@dmeza sorry I can't seem to get this working on my machines:

On High Sierra/Chrome Version 63.0.3239.132 I just get a blank browser page trying to open the mattermost permalink

![jan-24-2018 12-41-13](https://user-images.githubusercontent.com/14333569/35353301-01ff1d98-0104-11e8-9be9-e8842c381842.gif)

On Win 10/Chrome I get the pop up asking to open Mattermost but a blank browser page after that

esethna commented:
@dmeza anything I can help with get the above issue resolved?

jasonblais commented:
@dmeza Quick questions:
1) Has there been any recent discussion about this PR outside of this thread?
2) Did you have any issues when you tested the changes? Wondering why it's not working for Eric. If you didn't have issues, I can ask QA to test and see if they have any.

esethna commented:
@dmeza just checking in on this PR. Let us know if you are having trouble reproing the issue I ran into and we can get QA from our side involved in narrowing it down.

cpanato commented:
Removed the test server. If need a new one, please rebase and resolve the conflicts.

thanks for this PR.

dmeza commented:
@esethna @jasonblais @csduarte: sorry for the long delay on getting to fix this. It's rebased to the latest from master and refactored to use new router logic. All checks passed.

I tested on MacOS with Safari, Chrome and Firefox. On Windows with Chrome, Safari, Firefox and Edge 20 & 40. (IE 11 did not even load MM master home page).
I added a hack to redirect to desktop application in production mode (pre-release) to make testing easier.

The way to test it is to use the url of the test server and add `/vault#<<relative path in pre-release>>`. So for example if, the test server is this:
http://i-068bb459345e5ab31.test.spinmint.com
and you want to fire the native app to go to:
https://pre-release.mattermost.com/core/channels/apiv4
And then to:
https://pre-release.mattermost.com/core/channels/desktop-app

You would use: 
http://i-068bb459345e5ab31.test.spinmint.com/vault#/core/channels/apiv4
http://i-068bb459345e5ab31.test.spinmint.com/vault#/core/channels/desktop-app

We can have a call if the way it's set up is not clear.

The following gifs show you the captures from MacOS:
![mmlandingpage_chrome](https://user-images.githubusercontent.com/160621/41492062-3e20695c-70c2-11e8-8c3f-c42ff3025bf6.gif)
![mmlandingpage_firefox](https://user-images.githubusercontent.com/160621/41492063-3e41ca0c-70c2-11e8-81cc-fbc62c37c3a6.gif)
![mmlandingpage_safari](https://user-images.githubusercontent.com/160621/41492064-3e746764-70c2-11e8-98c2-36394122a2d2.gif)

dmeza commented:
@esethna @jasonblais: @csduarte proposes that we integrate further with the `/pl/` route, so that no further changes are needed on the mobile apps. Mobile apps right now handle deeplinking directly from the protocol and domain. That means that if you click on a link on an email they handle it directly. For desktop it's not possible, so we need the landing page.  If this makes sense, please let's set up a call to discuss the approach.

esethna commented:
Thanks @dmeza. What's involved in integrating with the /pl/ route? Is it further work on this PR or something to be submitted separately, and to which repros? If there's no more work involved in this PR we can move ahead with the testing.

dmeza commented:
@esethna yes to integrate with `/pl/` route would mean further work in this PR and no other changes to other repos. If we leave the PR as is there would need to be a change to the React Native mobile repo to handle the `/vault` url.

dmeza commented:
@esethna the idea is to agree on an approach so that there's no big changes later.

dmeza commented:
@yuya-oc rebased and addressed the changes requested. The only change not done is the sizes of the icons of the dmg in `electron-builder.json`. It's required to make the icon bigger and the font bigger also.
`
    "iconSize":** 140,
    "iconTextSize": 18
`

esethna commented:
@dmeza saw this PR on the mobile repo: https://github.com/mattermost/mattermost-mobile/pull/1805

How does that affect the work you're proposing on this PR? Are you still proposing integrating with /pl/ route?

dmeza commented:
@esethna they are related in that they add support to deeplinking into the mobile app, but it doesn't have anything to do with the landing page or any specific route.

dmeza commented:
@esethna rebased to latest from master and fixed the order or regexp for protocol to account correctly for https.

esethna commented:
@enahum have you had a chance to take a look at this PR and comment on the approach?

esethna commented:
Thanks @enahum. @dmeza let us know if you'd like to jump on a call to discuss further

dmeza commented:
@esethna @enahum yes, let's have a call tomorrow morning. Then I can show you how it works and answer any questions or concerns.

dmeza commented on package.json:
```diff
@@ -10,6 +10,7 @@
     "bootstrap-colorpicker": "2.5.2",
     "chart.js": "2.7.2",
     "compass-mixins": "0.12.10",
+    "custom-protocol-detection": "uber-uchat/custom-protocol-detection",
```
Yes, it's a fork and it adds `Safari support`:
 https://github.com/uber-uchat/custom-protocol-detection/commits/master

dmeza commented on components/go_to_native_app/go_to_native_app.jsx:
```diff
@@ -0,0 +1,107 @@
+            );
+        }
+
+        // prompt user to download in case they don't have the mobile app.
```
Yes, in case the user doesn't have it installed it should work for mobile and desktop apsp.

esethna commented:
@dmeza thanks for the call today. Action items we discussed would be:

1. Updating the links in email notifications to the vault route
2. Admin setting
3. Potential redesign of the landing page
4. Ability to save choice for opening webapp

I will follow up on the above, do some more testing and spec out the proposed changes after consulting with the design team.

esethna commented:
Just an update, design team is working on an updated version of the landing page currently ^

dmeza commented:
@esethna rebased to the laster from master.
